core/libp2p/protocol_muxer/multiselect/multiselect.cpp

```diff
@@ -0,0 +1,283 @@
+        // if the protocol is not available, send ls
+        sendLsMsg(std::move(stream_state));
+        return;
+      case Status::PROTOCOL_SENT:
+        // this is ack that the protocol we want to communicate over is
+        // supported by the other side; negotiation is finished
+        stream_state.proto_callback_(protocol);
+        return;
+      case Status::PROTOCOLS_SENT:
+        // the other side has chosen a protocol to communicate over; send an
+        // ack, and negotiation is finished
+        stream_state.proto_callback_(protocol);
+        sendProtocolMsg(protocol, true, false, std::move(stream_state));
+        return;
+      case Status::LS_SENT:
+        // we asked for a list of supported protocols, and the other side
+        // responded with only one; if it's not available on our side, send na?;
+        // otherwise, negotiation is finished
+        if (std::find(supported_protocols_.begin(), supported_protocols_.end(),
+                      protocol)
+            != supported_protocols_.end()) {
+          stream_state.proto_callback_(protocol);
+          sendProtocolMsg(protocol, false, false, std::move(stream_state));
+          return;
+        }
+        stream_state.proto_callback_(MultiselectErrors::NEGOTIATION_FAILED);
+        sendNaMsg(std::move(stream_state));
+        return;
+      case Status::NA_SENT:
+        // unexpected request-response combination; send ls
+        sendLsMsg(std::move(stream_state));
+        return;
+    }
+  }
+
+  void Multiselect::handleProtocolsMsg(const std::vector<Protocol> &protocols,
+                                       StreamState stream_state) const {
+    using Status = StreamState::NegotiationStatus;
+
+    switch (stream_state.status_) {
+      case Status::OPENING_SENT:
+      case Status::PROTOCOL_SENT:
+      case Status::PROTOCOLS_SENT:
+      case Status::NA_SENT:
+        // unexpected request-response combinations; send ls
+        sendLsMsg(std::move(stream_state));
+        return;
+      case Status::LS_SENT:
+        // if any of the received protocols is supported by our side, choose it;
+        // send NA otherwise?
+        for (const auto &proto : protocols) {
+          // as size of vectors should be around 10 or less, we can use O(n*n)
+          // approach
+          if (std::find(supported_protocols_.begin(),
+                        supported_protocols_.end(), proto)
+              != supported_protocols_.end()) {
+            // the protocol is found; send it as our choice
+            sendProtocolMsg(proto, false, true, std::move(stream_state));
+            return;
+          }
+        }
+        stream_state.proto_callback_(MultiselectErrors::NEGOTIATION_FAILED);
+        sendNaMsg(std::move(stream_state));
+        return;
+    }
+  }
+
+  void Multiselect::handleLsMsg(StreamState stream_state) const {
+    // respond with a list of protocols, supported by us
+    sendProtocolsMsg(gsl::span<const Protocol>(supported_protocols_),
```

Comment:
do not cast to gsl::span, vector can be used directly

Reply:
also, use ProtocolRepository to get OUR supported protocols. If you need better interface for it - either write it, or ask me to do it